lambdas/thumbnail/index.py

```diff
@@ -0,0 +1,92 @@
+"""
+Generate thumbnails for images in S3.
+"""
+import base64
+from io import BytesIO
+import json
+import os
+
+from PIL import Image
+import requests
+
+from t4_lambda_shared.decorator import api, validate
+
+
+# Eventually we'll want to precompute/cache thumbnails, so we won't be able to support
+# arbitrary sizes. Might as well copy Dropbox' API:
+# https://www.dropbox.com/developers/documentation/http/documentation#files-get_thumbnail
+SUPPORTED_SIZES = [
+    (32, 32),
+    (64, 64),
+    (128, 128),
+    (256, 256),
+    (480, 320),
+    (640, 480),
+    (960, 640),
+    (1024, 768),
+    (2048, 1536)
+]
+# Map URL parameters to actual sizes, e.g. 'w128h128' -> (128, 128)
+# TODO(dima): Is this a good idea? Don't know why Dropbox does it that way.
+SIZE_PARAMETER_MAP = {f'w{w}h{h}': (w, h) for w, h in SUPPORTED_SIZES}
+
+ALLOWED_ORIGINS = [
+    'http://localhost:3000',
+    os.environ.get('WEB_ORIGIN')
+]
+
+SCHEMA = {
+    'type': 'object',
+    'properties': {
+        'url': {
+            'type': 'string'
+        },
+        'size': {
+            'enum': list(SIZE_PARAMETER_MAP)
+        }
+    },
+    'required': ['url', 'size'],
+    'additionalProperties': False
+}
+
+@api(cors_origins=ALLOWED_ORIGINS)
+@validate(SCHEMA)
```

Comment:
If the client passes a bad size this will 400... that's fine but it would be nice to have a message: "Unsupported size {whatever} requested. Supported sizes are {whatever}."

Reply:
Actually, the JSON schema error message is pretty helpful (if a bit verbose). E.g., passing "foo" results in:
```
'foo' is not one of ['w32h32', 'w64h64', 'w128h128', 'w256h256', 'w480h320', 'w640h480', 'w960h640', 'w1024h768', 'w2048h1536']

Failed validating 'enum' in schema['properties']['size']:
    {'enum': ['w32h32',
              'w64h64',
              'w128h128',
              'w256h256',
              'w480h320',
              'w640h480',
              'w960h640',
              'w1024h768',
              'w2048h1536']}

On instance['size']:
    'foo'
```

Reply:
Oh nice I should have noticed that